File: evals/rubric.go
package evals

import (
	"context"
	"fmt"

	"github.com/openai/openai-go/v3"
)

// Rubric defines an evaluation criterion with a name, description, and grader.
type Rubric struct {
	Name        string
	Description string
	Grader      Grader
}

// ConfidenceInterval represents a statistical confidence interval.
type ConfidenceInterval struct {
	Lower float64
	Upper float64
	Level float64
}

// Score holds the result of evaluating output against a single rubric.
type Score struct {
	RubricName  string
	Passed      bool
	Value       int
	Reasoning   string
	GraderType  string
	GraderModel string
	PassRate    float64
	Samples     int

	StandardDeviation  float64
	Variance           float64
	ConfidenceInterval ConfidenceInterval
	PassCount          int
	FailCount          int
}

// Grader evaluates output against a rubric criterion.
type Grader interface {
	Grade(ctx context.Context, client *openai.Client, model string, rubric Rubric, output string) Score
}

// GraderWithInputs is an extended grader that can access input variables.
type GraderWithInputs interface {
	Grader
	GradeWithInputs(ctx context.Context, client *openai.Client, model string, rubric Rubric, output string, inputs map[string]any) Score
}

// LLMGrader uses a language model to evaluate output.
// Set Samples > 1 to run multiple grading samples and aggregate results for better stability.
type LLMGrader struct {
	Samples int
}

// Grade evaluates the output using an LLM judge with structured output.
// When Samples > 1, it generates multiple grading responses and aggregates them using mean pass rate.
func (g LLMGrader) Grade(ctx context.Context, client *openai.Client, model string, rubric Rubric, output string) Score {
	return g.GradeWithInputs(ctx, client, model, rubric, output, nil)
}

// GradeWithInputs evaluates the output using an LLM judge with access to input variables.
func (g LLMGrader) GradeWithInputs(ctx context.Context, client *openai.Client, model string, rubric Rubric, output string, inputs map[string]any) Score {
	prompt := g.buildPrompt(rubric, output, inputs)
	return g.gradeWithPrompt(ctx, client, model, rubric, prompt)
}

func (g LLMGrader) buildPrompt(rubric Rubric, output string, inputs map[string]any) string {
	inputsStr := ""
	if len(inputs) > 0 {
		inputsStr = "\nInput Variables:\n"
		for k, v := range inputs {
			inputsStr += fmt.Sprintf("- %s: %v\n", k, v)
		}
	}

	return fmt.Sprintf(`You are an evaluation judge. Score the following output against this rubric.

Rubric: %s
Description: %s
%s
Output to evaluate:
---
%s
---`, rubric.Name, rubric.Description, inputsStr, output)
}

func (g LLMGrader) gradeWithPrompt(ctx context.Context, client *openai.Client, model string, rubric Rubric, prompt string) Score {
	score := Score{
		RubricName:  rubric.Name,
		GraderType:  "llm",
		GraderModel: model,
		Samples:     1,
	}

	n := g.Samples
	if n <= 1 {
		graderResp, err := generateStructuredGraderResponse(ctx, client, model, prompt)
		if err != nil {
			score.Reasoning = fmt.Sprintf("grader error: %v", err)
			return score
		}

		score.Passed = graderResp.Verdict == "PASS"
		if score.Passed {
			score.Value = 1
		}
		score.Reasoning = graderResp.Reasoning
		score.PassRate = 1.0
		if !score.Passed {
			score.PassRate = 0.0
		}
		return score
	}

	responses, err := generateStructuredGraderResponseMulti(ctx, client, model, prompt, n)
	if err != nil {
		score.Reasoning = fmt.Sprintf("grader error: %v", err)
		return score
	}

	agg := aggregateGraderResponses(responses)
	score.Passed = agg.Passed
	if score.Passed {
		score.Value = 1
	}
	score.Reasoning = agg.Reasoning
	score.PassRate = agg.PassRate
	score.Samples = agg.Samples
	score.StandardDeviation = agg.StandardDeviation
	score.Variance = agg.Variance
	score.ConfidenceInterval = agg.ConfidenceInterval
	score.PassCount = agg.PassCount
	score.FailCount = agg.FailCount

	return score
}

// FuncGrader wraps a function to implement the Grader interface.
type FuncGrader func(output string) (passed bool, reason string)

// Grade evaluates the output using the wrapped function.
func (g FuncGrader) Grade(_ context.Context, _ *openai.Client, _ string, rubric Rubric, output string) Score {
	score := Score{
		RubricName: rubric.Name,
		GraderType: "func",
	}

	if g == nil {
		score.Reasoning = "no grader function provided"
		return score
	}

	passed, reasoning := g(output)
	score.Passed = passed
	if passed {
		score.Value = 1
	}
	score.Reasoning = reasoning

	return score
}

// FuncGraderWithInputs wraps a function that needs access to input variables.
type FuncGraderWithInputs func(output string, inputs map[string]any) (passed bool, reason string)

// Grade implements the base Grader interface by calling with nil inputs.
func (g FuncGraderWithInputs) Grade(ctx context.Context, client *openai.Client, model string, rubric Rubric, output string) Score {
	return g.GradeWithInputs(ctx, client, model, rubric, output, nil)
}

// GradeWithInputs evaluates the output using the wrapped function with inputs.
func (g FuncGraderWithInputs) GradeWithInputs(_ context.Context, _ *openai.Client, _ string, rubric Rubric, output string, inputs map[string]any) Score {
	score := Score{
		RubricName: rubric.Name,
		GraderType: "func",
	}

	if g == nil {
		score.Reasoning = "no grader function provided"
		return score
	}

	passed, reasoning := g(output, inputs)
	score.Passed = passed
	if passed {
		score.Value = 1
	}
	score.Reasoning = reasoning

	return score
}

File: crates/eval/src/rubric.rs
use std::collections::HashMap;

use crate::{
    ChatCompleter, ConfidenceInterval, aggregate_grader_responses,
    generate_structured_grader_response, generate_structured_grader_response_multi,
};

/// Score represents the result of evaluating output against a rubric.
#[derive(Debug, Clone, Default)]
pub struct Score {
    pub rubric_name: String,
    pub passed: bool,
    pub value: i32,
    pub reasoning: String,
    pub grader_type: String,
    pub grader_model: String,
    pub pass_rate: f64,
    pub samples: i32,
    pub standard_deviation: f64,
    pub variance: f64,
    pub confidence_interval: ConfidenceInterval,
    pub pass_count: i32,
    pub fail_count: i32,
}

/// Grades output using an LLM-based grader.
///
/// # Arguments
/// * `client` - The chat completion client
/// * `model` - The model to use for grading
/// * `rubric_name` - Name of the rubric
/// * `rubric_description` - Description of what the rubric evaluates
/// * `samples` - Number of grading samples for consensus (1 for single evaluation)
/// * `output` - The output to evaluate
/// * `inputs` - Optional input variables for context
/// * `on_evaluation` - Optional callback called after each evaluation
#[allow(clippy::too_many_arguments)]
pub fn grade_with_llm(
    client: &dyn ChatCompleter,
    model: &str,
    rubric_name: &str,
    rubric_description: &str,
    samples: i32,
    output: &str,
    inputs: Option<&HashMap<String, serde_json::Value>>,
    on_evaluation: Option<&dyn Fn()>,
) -> Score {
    let samples = if samples <= 0 { 1 } else { samples };
    let prompt = build_llm_prompt(rubric_name, rubric_description, output, inputs);

    let mut score = Score {
        rubric_name: rubric_name.to_string(),
        grader_type: "llm".to_string(),
        grader_model: model.to_string(),
        samples: 1,
        ..Default::default()
    };

    if samples <= 1 {
        match generate_structured_grader_response(client, model, &prompt) {
            Ok(grader_resp) => {
                score.passed = grader_resp.verdict == "PASS";
                score.value = if score.passed { 1 } else { 0 };
                score.reasoning = grader_resp.reasoning;
                score.pass_rate = if score.passed { 1.0 } else { 0.0 };
            }
            Err(e) => {
                score.reasoning = format!("grader error: {}", e);
            }
        }
        if let Some(cb) = on_evaluation {
            cb();
        }
        return score;
    }

    match generate_structured_grader_response_multi(client, model, &prompt, samples) {
        Ok(responses) => {
            if let Some(cb) = on_evaluation {
                for _ in 0..samples {
                    cb();
                }
            }

            let agg = aggregate_grader_responses(&responses);
            score.passed = agg.passed;
            score.value = if score.passed { 1 } else { 0 };
            score.reasoning = agg.reasoning;
            score.pass_rate = agg.pass_stats.pass_rate;
            score.samples = agg.pass_stats.samples;
            score.standard_deviation = agg.pass_stats.standard_deviation;
            score.variance = agg.pass_stats.variance;
            score.confidence_interval = agg.pass_stats.confidence_interval;
            score.pass_count = agg.pass_stats.pass_count;
            score.fail_count = agg.pass_stats.fail_count;
        }
        Err(e) => {
            score.reasoning = format!("grader error: {}", e);
        }
    }

    score
}

/// Grades output using a function-based grader.
pub fn grade_with_func(
    rubric_name: &str,
    output: &str,
    grader_fn: fn(&str) -> (bool, String),
) -> Score {
    let (passed, reasoning) = grader_fn(output);
    Score {
        rubric_name: rubric_name.to_string(),
        passed,
        value: if passed { 1 } else { 0 },
        reasoning,
        grader_type: "func".to_string(),
        ..Default::default()
    }
}

fn build_llm_prompt(
    rubric_name: &str,
    rubric_description: &str,
    output: &str,
    inputs: Option<&HashMap<String, serde_json::Value>>,
) -> String {
    let mut inputs_str = String::new();
    if let Some(inputs) = inputs {
        if !inputs.is_empty() {
            inputs_str.push_str("\nInput Variables:\n");
            for (k, v) in inputs {
                inputs_str.push_str(&format!("- {}: {}\n", k, v));
            }
        }
    }

    format!(
        r#"You are an evaluation judge. Score the following output against this rubric.

Rubric: {}
Description: {}
{}
Output to evaluate:
---
{}
---"#,
        rubric_name, rubric_description, inputs_str, output
    )
}

/// A simple grader function that checks if output is non-empty.
pub fn is_non_empty(output: &str) -> (bool, String) {
    if output.trim().is_empty() {
        (false, "output is empty".to_string())
    } else {
        (true, "output is non-empty".to_string())
    }
}
